File: frontend/src/app/services/api.service.ts
import { Injectable, signal } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class ApiService {
  private baseUrl = 'https://trading-journal.fastapicloud.dev';
  
  // Active User ID signal - allows mock switcher in header to trigger reactive reloading
  // Active User ID signal - allows mock switcher in header to trigger reactive reloading
  readonly activeUserId = signal<string>('u001');

  // Google authentication profile signal
  readonly googleProfile = signal<any | null>(null);

  constructor(private http: HttpClient) {
    // Automatically restore cached session on instantiation
    const cachedProfile = localStorage.getItem('google_profile');
    if (cachedProfile) {
      try {
        const profile = JSON.parse(cachedProfile);
        this.googleProfile.set(profile);
        this.activeUserId.set(profile.userId);
      } catch (e) {
        localStorage.removeItem('google_profile');
        this.activeUserId.set('guest');
      }
    } else {
      this.activeUserId.set('guest');
    }
  }

  isGuest(): boolean {
    return this.googleProfile() === null;
  }

  logout() {
    localStorage.removeItem('google_profile');
    this.googleProfile.set(null);
    this.activeUserId.set('guest');
  }

  loginAsDemo(userId: string, name: string) {
    const profile = { userId, name, email: `${userId}@demo.com` };
    localStorage.setItem('google_profile', JSON.stringify(profile));
    this.googleProfile.set(profile);
    this.activeUserId.set(userId);
  }

  private getHeaders(): HttpHeaders {
    return new HttpHeaders({
      'X-User-Id': this.activeUserId()
    });
  }

  getPrices(): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/api/prices`);
  }

  getPrice(symbol: string): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/api/price/${symbol}`);
  }

  getTrades(ticker?: string): Observable<any[]> {
    const url = ticker ? `${this.baseUrl}/api/trades?ticker=${ticker}` : `${this.baseUrl}/api/trades`;
    return this.http.get<any[]>(url, { headers: this.getHeaders() });
  }

  addTrade(trade: { ticker: string; side: string; price: number; qty: number }): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/api/trades`, trade, { headers: this.getHeaders() });
  }

  initTrade(trade: { ticker: string; side: string; price: number; qty: number }): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/api/trades/init`, trade, { headers: this.getHeaders() });
  }

  confirmTrade(trade: { ticker: string; side: string; price: number; qty: number }): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/api/trades/confirm`, trade, { headers: this.getHeaders() });
  }

  getPosition(symbol: string): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/api/position/${symbol}`, { headers: this.getHeaders() });
  }

  getPortfolio(): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/api/portfolio`, { headers: this.getHeaders() });
  }

  getTopOrders(): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/api/top-orders`, { headers: this.getHeaders() });
  }

  getTopTickers(): Observable<any[]> {
    return this.http.get<any[]>(`${this.baseUrl}/api/top-tickers`, { headers: this.getHeaders() });
  }

  getChartUrl(symbol: string): string {
    return `${this.baseUrl}/api/chart/${symbol}?userId=${this.activeUserId()}`;
  }

  googleLogin(credential: string): Observable<any> {
    return this.http.post<any>(`${this.baseUrl}/api/auth/google`, { credential });
  }
}

File: app/lib/services/api_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class ApiService {
  // Singleton instance
  static final ApiService instance = ApiService._internal();
  ApiService._internal();

  String get baseUrl => 'https://trading-journal.fastapicloud.dev';

  // Active user ID: 'guest' until a Google (or demo) sign-in sets it, mirroring
  // the web app's ApiService.activeUserId default (frontend/src/app/services/api.service.ts).
  String activeUserId = 'guest';

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        'X-User-Id': activeUserId,
      };

  Future<List<dynamic>> getPrices() async {
    final response = await http.get(Uri.parse('$baseUrl/api/prices'));
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    }
    throw Exception('Failed to load prices');
  }

  Future<List<dynamic>> getTrades({String? ticker}) async {
    String url = '$baseUrl/api/trades';
    if (ticker != null && ticker.isNotEmpty) {
      url += '?ticker=${ticker.toUpperCase()}';
    }
    final response = await http.get(Uri.parse(url), headers: _headers);
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    }
    throw Exception('Failed to load trades');
  }

  Future<Map<String, dynamic>> addTrade(String ticker, String side, int price, int qty) async {
    final body = jsonEncode({
      'ticker': ticker.toUpperCase(),
      'side': side.toUpperCase(),
      'price': price,
      'qty': qty,
    });
    final response = await http.post(
      Uri.parse('$baseUrl/api/trades'),
      headers: _headers,
      body: body,
    );
    return jsonDecode(response.body);
  }

  Future<Map<String, dynamic>> initTrade(String ticker, String side, int price, int qty) async {
    final body = jsonEncode({
      'ticker': ticker.toUpperCase(),
      'side': side.toUpperCase(),
      'price': price,
      'qty': qty,
    });
    final response = await http.post(
      Uri.parse('$baseUrl/api/trades/init'),
      headers: _headers,
      body: body,
    );
    final decoded = jsonDecode(response.body);
    if (response.statusCode >= 400) {
      throw Exception(decoded['detail'] ?? decoded['error'] ?? 'Failed to validate trade');
    }
    return decoded;
  }

  Future<Map<String, dynamic>> confirmTrade(String ticker, String side, int price, int qty) async {
    final body = jsonEncode({
      'ticker': ticker.toUpperCase(),
      'side': side.toUpperCase(),
      'price': price,
      'qty': qty,
    });
    final response = await http.post(
      Uri.parse('$baseUrl/api/trades/confirm'),
      headers: _headers,
      body: body,
    );
    final decoded = jsonDecode(response.body);
    if (response.statusCode >= 400) {
      throw Exception(decoded['detail'] ?? decoded['error'] ?? 'Failed to submit trade');
    }
    return decoded;
  }

  Future<Map<String, dynamic>> getPosition(String symbol) async {
    final response = await http.get(
      Uri.parse('$baseUrl/api/position/${symbol.toUpperCase()}'),
      headers: _headers,
    );
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    }
    throw Exception('Failed to load position details');
  }

  Future<List<dynamic>> getPortfolio() async {
    final response = await http.get(Uri.parse('$baseUrl/api/portfolio'), headers: _headers);
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    }
    throw Exception('Failed to load portfolio');
  }

  Future<List<dynamic>> getTopOrders() async {
    final response = await http.get(Uri.parse('$baseUrl/api/top-orders'), headers: _headers);
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    }
    throw Exception('Failed to load top orders');
  }

  Future<List<dynamic>> getTopTickers() async {
    final response = await http.get(Uri.parse('$baseUrl/api/top-tickers'), headers: _headers);
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    }
    throw Exception('Failed to load top tickers');
  }

  Future<Map<String, dynamic>> googleLogin(String credential) async {
    final response = await http.post(
      Uri.parse('$baseUrl/api/auth/google'),
      headers: const {'Content-Type': 'application/json'},
      body: jsonEncode({'credential': credential}),
    );
    final decoded = jsonDecode(response.body);
    if (response.statusCode >= 400) {
      throw Exception(decoded['detail'] ?? 'Google sign-in failed');
    }
    return decoded;
  }
}
